### PR TITLE
[Explore Vis]Simplify threshold logic

### PR DESCRIPTION
**Pull request by Qxisylolo**
### Description
this pr simplifies threshold logic

## Changelog
- feat: Simplify threshold logic

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

## Summary by CodeRabbit

* **Bug Fixes**
  * Enhanced threshold handling and color consistency across gauge and metric visualizations.
  * Improved support for undefined values and edge cases in threshold processing.

* **Tests**
  * Added comprehensive unit tests for threshold generation, filtering, range handling, and deduplication.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/plugins/explore/public/components/visualizations/metric/to_expression.ts [hunk 1]
```diff
@@ -63,24 +60,23 @@ export const createSingleMetric = (
...
-
   function targetFillColor(
     useThresholdColor: boolean,
     threshold?: Threshold[],
     baseColor?: string
   ) {
     const newThreshold = threshold ?? [];
 
-    const newBaseColor = baseColor ?? getColors().statusGreen;
+    if (calculatedValue === undefined || calculatedValue < minBase) {
```
Comment (ruanyl) [inline src/plugins/explore/public/components/visualizations/metric/to_expression.ts]: It seems we don't consider `maxBase`, I'm just thinking shall we also not consider `minBase` to make it consistent? so just make the text color solely based on thresholds.

File: src/plugins/explore/public/components/visualizations/bar_gauge/bar_gauge_utils.ts [hunk 2]
```diff
@@ -157,3 +142,68 @@ export const generateParams = (
...
+  }
+
+  if (lastBelowIndex !== -1) {
+    result.splice(0, lastBelowIndex);
+    result[0] = { ...result[0], value: minBase };
+  } else {
+    result.unshift({ value: minBase, color: defaultColor });
+  }
+
+  const valueResults: Threshold[] = [];
```
Comment (ruanyl) [inline src/plugins/explore/public/components/visualizations/bar_gauge/bar_gauge_utils.ts]: I would create another function to calculate the `valueResults` separately

**Review by ruanyl — DISMISSED** [review]

File: src/plugins/explore/public/components/visualizations/metric/to_expression.ts (hunk 1 above)
Comment (Qxisylolo) [reply src/plugins/explore/public/components/visualizations/metric/to_expression.ts]: sure, updated

File: src/plugins/explore/public/components/visualizations/bar_gauge/bar_gauge_utils.ts [hunk 3]
```diff
@@ -157,3 +142,75 @@ export const generateParams = (
...
+export const generateThresholds = (
+  minBase: number,
+  maxBase: number,
+  thresholds: Threshold[],
+  baseColor: string | undefined
+) => {
+  const defaultColor = baseColor ?? getColors().statusGreen;
+
+  // sort thresholds by value and dedupe threshold by value
+  thresholds = thresholds
```
Comment (wanglam) [inline src/plugins/explore/public/components/visualizations/bar_gauge/bar_gauge_utils.ts:155]: Nit: can be refactor with code below:
```tsx
thresholds = Array.from(
  thresholds.reduce((map, threshold) => {
    map.set(threshold.value, threshold);
    return map;
  }, new Map<number, Threshold>()).values()
)
.sort((t1, t2) => t1.value - t2.value);
```
It seems more clear to me.

File: src/plugins/explore/public/components/visualizations/bar_gauge/bar_gauge_utils.ts [hunk 4]
```diff
@@ -157,3 +142,75 @@ export const generateParams = (
...
+      while (
+        thresholdIndex < thresholds.length - 1 &&
+        thresholds[thresholdIndex + 1].value <= stop
+      ) {
+        thresholdIndex++;
+      }
+
+      // Add valid threshold for this stop
+      if (thresholds[thresholdIndex].value <= stop) {
+        valueThresholds.push({ value: stop, color: thresholds[thresholdIndex].color });
```
Comment (wanglam) [inline src/plugins/explore/public/components/visualizations/bar_gauge/bar_gauge_utils.ts:211]: Do we need to return here if thresholdIndex equal `thresholds.length - 1`? 
Imagine the input parameters like below:
minBase = 0; maxBase = 100; valueStops = [10,20,30,40,50]; thresholds = [{ value: 15, color: "green" }];
The result will be:
```
[{"value":20,"color":"green"},{"value":30,"color":"green"},{"value":40,"color":"green"},{"value":50,"color":"green"}]
```
I'm not sure if it's meet our requirements.

Comment (Qxisylolo) [reply src/plugins/explore/public/components/visualizations/bar_gauge/bar_gauge_utils.ts:211]: Yes, this is intended, if all the value stops are beyond the last threshold, then we will use that threshold to create value thresholds, so that, the gradient mode in different bar will be consist.
